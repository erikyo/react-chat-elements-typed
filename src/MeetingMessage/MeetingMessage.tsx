import type React from "react";
import { type FC, type MouseEvent, useState } from "react";
import "./MeetingMessage.css";

import { FaCalendar, FaCaretDown, FaCaretRight } from "react-icons/fa";
import { HiOutlineVideoCamera } from "react-icons/hi";
import { IoMdChatboxes } from "react-icons/io";
import { MdMoreHoriz } from "react-icons/md";

import { format } from "date-fns";

import Avatar from "../Avatar/Avatar";
import Dropdown from "../Dropdown/Dropdown";

import classNames from "classnames";
import type { IMeetingMessageProps } from "../types";
import { DATE_FORMAT } from "../constants";
import Button from "../Button/Button";

const MeetingMessage: FC<IMeetingMessageProps> = (props) => {
	const [toggle, setToggle] = useState(false);

	const PARTICIPANT_LIMIT: number =
		props.participantsLimit || Number.POSITIVE_INFINITY;

	const dateText = props.dateString
		? props.dateString
		: props.date && format(props.date, DATE_FORMAT);

	return (
		<div className="rce-mbox-mtmg">
			<div className="rce-mtmg">
				<div className="rce-mtmg-subject">
					{props.subject || "Unknown Meeting"}
				</div>
				<div className="rce-mtmg-body">
					<div className="rce-mtmg-item">
						<FaCalendar />
						<div className="rce-mtmg-content">
							<span className="rce-mtmg-title">{props.title}</span>
							<span className="rce-mtmg-date">{dateText}</span>
						</div>
					</div>

					{props.onMeetingMoreSelect &&
						props.moreItems &&
						props.moreItems.length > 0 && (
							<div>
								<Dropdown
									animationType="bottom"
									animationPosition="northeast"
									buttonProps={{
										className: "rce-mtmg-right-icon",
										icon: {
											component: <MdMoreHoriz />,
											size: 24,
										},
									}}
									items={props.moreItems}
									onSelect={(e) => props.onMeetingMoreSelect?.(e, props)}
								/>
							</div>
						)}
				</div>
				<Button
					className="rce-mtmg-body-bottom"
					link
					onClick={() => setToggle(!toggle)}
					icon={{
						component: toggle ? <FaCaretDown /> : <FaCaretRight />,
					}}
					text={
						toggle ? (
							<div className="rce-mtmg-bottom--tptitle">
								<span>{props.collapseTitle}</span>
							</div>
						) : (
							<div className="rce-mtmg-body-bottom--bttitle">
								<span>
									{props.participants
										?.slice(0, PARTICIPANT_LIMIT)
										.map((x) => x.title || "Unknown")
										.join(", ")}
									{props.participants &&
										PARTICIPANT_LIMIT &&
										props.participants.length > PARTICIPANT_LIMIT &&
										`, +${props.participants.length - PARTICIPANT_LIMIT}`}
								</span>
							</div>
						)
					}
				/>
				<div
					className={classNames("rce-mtmg-toogleContent", {
						"rce-mtmg-toogleContent--click": toggle,
					})}
				>
					{props.dataSource?.map((x, i) => {
						return (
							<div key={x.id ?? i.toString()}>
								{!x.event && (
									<div className="rce-mitem">
										<div
											className={classNames("rce-mitem avatar", {
												"rce-mitem no-avatar": !x.avatar,
											})}
										>
											{x.avatar ? (
												<Avatar title={x.title || "Unknown"} src={x.avatar} />
											) : (
												<IoMdChatboxes />
											)}
										</div>
										<div className="rce-mitem-body">
											<div className="rce-mitem-body--top">
												<div
													className="rce-mitem-body--top-title"
													onClick={(e) => props.onMeetingLink?.(e, props)}
													onKeyDown={console.log}
												>
													{x.title}
												</div>
												<div className="rce-mitem-body--top-time">
													{x.dateString
														? x.dateString
														: x.date && x.date && format(x.date, DATE_FORMAT)}
												</div>
											</div>
											<div className="rce-mitem-body--bottom">
												<div className="rce-mitem-body--bottom-title">
													{x.message}
												</div>
											</div>
										</div>
									</div>
								)}
								{x.event && (
									<div className="rce-mitem-event">
										<div className="rce-mitem-bottom-body">
											<div className="rce-mitem-event-wrapper">
												<div className="rce-mitem-body avatar">
													<HiOutlineVideoCamera />
												</div>
												<div className="rce-mitem-bottom-body-top">
													{x.event.title}
													<div className="rce-mitem-body--top-time">
														{x.dateString
															? x.dateString
															: x.date && format(x.date, "yyyy-MM-dd HH:mm")}
													</div>
													<div className="rce-mitem-avatar-content">
														{
															<div className="rce-mitem-avatar">
																{x.event.avatars
																	?.slice(0, PARTICIPANT_LIMIT)
																	.map((x, i) => (
																		<Avatar
																			title={x.title}
																			key={i.toString()}
																			src={x.src}
																			size={"small"}
																		/>
																	))}
																{x.event.avatars &&
																	x.event.avatarsLimit &&
																	x.event.avatars.length >
																		x.event.avatarsLimit && (
																		<div
																			className="rce-mitem-length rce-mitem-tooltip"
																			tooltip={x.event.avatars
																				.slice(
																					x.event.avatarsLimit,
																					x.event.avatars.length,
																				)
																				.map((avatar) => avatar.title)
																				.join(",")
																				.toString()}
																		>
																			<span className="rce-mitem-tooltip-text">
																				{`+${
																					x.event.avatars.length -
																					x.event.avatarsLimit
																				}`}
																			</span>
																		</div>
																	)}
															</div>
														}
													</div>
												</div>
											</div>
											{x.record && (
												<div className="rce-mtmg-call-record">
													<div className="rce-mtmg-call-body">
														<div
															onClick={(e) =>
																props.onMeetingVideoLink?.(e, props)
															}
															onKeyDown={(e) =>
																props.onMeetingVideoLink?.(e, props)
															}
															className="rce-mtmg-call-avatars"
														>
															<img
																alt="Meeting poster"
																className={"rce-mtmg-call-avatars"}
																src={x.record.cover}
															/>
															<div className={"rce-mtmg-record-time"}>
																{x.record.time}
															</div>
														</div>
														<div className="rce-mtmg-call-body-title">
															<span>{x.record.title}</span>
															<div className="rce-mtmg-call-body-bottom">
																{x.record.savedBy}
															</div>
														</div>
													</div>
												</div>
											)}
										</div>
									</div>
								)}
							</div>
						);
					})}
				</div>
			</div>
		</div>
	);
};

export default MeetingMessage;
