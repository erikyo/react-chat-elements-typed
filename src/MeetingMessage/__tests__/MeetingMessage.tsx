import { describe, expect, it } from "vitest";
import React from "react";
import MeetingMessage from "../MeetingMessage";
import { render } from "@testing-library/react";

describe("Avatar component", () => {
	it("should render without issues", () => {
		const component = render(
			<MeetingMessage type={"text"} date={new Date("2050-01-01")} />,
		);
		expect(component.container).toBeTruthy();
		expect(component.container).toMatchSnapshot();
	});
});
