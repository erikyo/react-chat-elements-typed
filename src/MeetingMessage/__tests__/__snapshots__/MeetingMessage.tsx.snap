// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar component > should render without issues 1`] = `
<div>
  <div
    class="rce-mbox-mtmg"
  >
    <div
      class="rce-mtmg"
    >
      <div
        class="rce-mtmg-subject"
      >
        Unknown Meeting
      </div>
      <div
        class="rce-mtmg-body"
      >
        <div
          class="rce-mtmg-item"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 192h424c6.6 0 12 5.4 12 12v260c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V204c0-6.6 5.4-12 12-12zm436-44v-36c0-26.5-21.5-48-48-48h-48V12c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v52H160V12c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v52H48C21.5 64 0 85.5 0 112v36c0 6.6 5.4 12 12 12h424c6.6 0 12-5.4 12-12z"
            />
          </svg>
          <div
            class="rce-mtmg-content"
          >
            <span
              class="rce-mtmg-title"
            />
            <span
              class="rce-mtmg-date"
            />
          </div>
        </div>
      </div>
      <button
        class="rce-button rce-mtmg-body-bottom link"
        style="border-width: 2px; flex-direction: row;"
        type="button"
      >
        <span>
          <div
            class="rce-mtmg-body-bottom--bttitle"
          >
            <span />
          </div>
        </span>
        <span
          class="rce-button-icon--container"
        >
          <span
            class="rce-button-icon"
            style="float: inherit; font-size: 12px;"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 192 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
              />
            </svg>
          </span>
        </span>
      </button>
      <div
        class="rce-mtmg-toogleContent"
      />
    </div>
  </div>
</div>
`;
